### PR TITLE
add beam demo

### PR DESCRIPTION
**Pull request by roywei**
Add beam demo with CTR prediction

### CODE REVIEW

File: apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ctr_prediction;
```
Comment (frankfliu) [inline apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java:13]: Please move to org.example.ctrprediction package

File: apache-beam/ctr_prediction/pom.xml [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+        <dependency>
+          <groupId>com.fasterxml.jackson.module</groupId>
+          <artifactId>jackson-module-scala_2.11</artifactId>
+          <version>${jackson.version}</version>
+          <scope>runtime</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+    <profile>
+      <id>samza-runner</id>
```
Comment (stu1130) [inline apache-beam/ctr_prediction/pom.xml:249]: do we need the samza here?

File: apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class CtrPrediction {
+
+    static class FeatureMap extends DoFn<String, String> {
```
Comment (frankfliu) [inline apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java:54]: Inner class should at the bottom of the file

File: apache-beam/ctr_prediction/pom.xml [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+          <groupId>org.apache.beam</groupId>
+          <artifactId>beam-runners-portability-java</artifactId>
+          <version>${beam.version}</version>
+          <scope>runtime</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+
+    <profile>
+      <id>dataflow-runner</id>
```
Comment (stu1130) [inline apache-beam/ctr_prediction/pom.xml:180]: Is this for GCP DataFlow?

File: apache-beam/ctr_prediction/pom.xml [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        <dependency>
+          <groupId>org.apache.beam</groupId>
+          <artifactId>beam-runners-samza</artifactId>
+          <version>${beam.version}</version>
+          <scope>runtime</scope>
+        </dependency>
+      </dependencies>
+    </profile>
+    <profile>
+      <id>nemo-runner</id>
```
Comment (stu1130) [inline apache-beam/ctr_prediction/pom.xml:260]: do we need the apache nemo here?

File: apache-beam/ctr_prediction/pom.xml [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+              <groupId>org.slf4j</groupId>
+              <artifactId>slf4j-log4j12</artifactId>
+            </exclusion>
+          </exclusions>
+        </dependency>
+      </dependencies>
+    </profile>
+
+    <profile>
+      <id>jet-runner</id>
```
Comment (stu1130) [inline apache-beam/ctr_prediction/pom.xml:286]: same as above

File: apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            for (int i = 1; i < fields.length; i++) {
+                features[i - 1] = Float.parseFloat(fields[i]);
+            }
+            return new NDList(manager.create(features));
+        }
+
+        @Override
+        public String processOutput(TranslatorContext ctx, NDList list) {
+            NDArray ctr = list.singletonOrThrow();
+            return String.valueOf(ctr.toFloatArray()[0]);
```
Comment (stu1130) [inline apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java:141]: If ctr is scalar, you can use
```suggestion
            return String.valueOf(ctr.getFloat());
```

File: apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        }));
+
+        // convert categorical features to integer according to feature map used in training
+        PCollection<String> preprocess = addIds.apply("Preprocess", ParDo.of(new FeatureMap()));
+
+        // run inference using Deep Java Library
+        PCollection<String> ctr = preprocess.apply("Inference", ParDo.of(new DoFn<String, String>() {
+            @ProcessElement
+            public void processElement(ProcessContext c) throws TranslateException, MalformedModelException, ModelNotFoundException, IOException {
+                Criteria<String, String> criteria = Criteria.builder()
```
Comment (frankfliu) [inline apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java:97]: We are loading the model on each process, we should move this to class member

File: apache-beam/ctr_prediction/.gitignore [hunk 9]
```diff
@@ -0,0 +1,5 @@
+.DS_Store
+.idea
+*.iml
```
Comment (lanking520) [inline apache-beam/ctr_prediction/.gitignore:3]: These should already be covered from the main ignore

File: apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java (hunk 8 above)
Comment (roywei) [reply apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java]: Our predictor is not serializable, I will see if I can update the DJL side

Comment (roywei) [reply apache-beam/ctr_prediction/src/main/java/ctr_prediction/CtrPrediction.java]: mvn, fixed by making it static

File: apache-beam/ctr_prediction/.gitignore (hunk 9 above)
Comment (roywei) [reply apache-beam/ctr_prediction/.gitignore]: removed as main ignore has them

**Pull request merged**